models/resnet.py

```diff
@@ -230,6 +236,137 @@ def forward(self, x):
         return logits, (layer0, maxpool, layer1, layer2, layer3, layer4, avgpool)
 
 
+class LinearWithFancyAttention(nn.Module):
```

Comment:
This looks cool, but does it need to be in the ResNet file? 

In general I think the main branch should be kept fairly minimal, and I think this goes beyond a minimal setup.

Also wouldn't this work for any model? Perhaps it would be better in its own file, passable to each of the models.